src/number_theory/lucas_lehmer.lean

```diff
@@ -0,0 +1,557 @@
+{ mul := λ x y, (x.1*y.1 + 3*x.2*y.2, x.1*y.2 + x.2*y.1) }
+
+@[simp] lemma mul_fst (x y : X q) : (x * y).1 = x.1 * y.1 + 3 * x.2 * y.2 := rfl
+@[simp] lemma mul_snd (x y : X q) : (x * y).2 = x.1 * y.2 + x.2 * y.1 := rfl
+
+instance : has_one (X q) :=
+{ one := ⟨1,0⟩ }
+
+@[simp] lemma one_fst : (1 : X q).1 = 1 := rfl
+@[simp] lemma one_snd : (1 : X q).2 = 0 := rfl
+
+@[simp] lemma bit0_fst (x : X q) : (bit0 x).1 = bit0 x.1 := rfl
+@[simp] lemma bit0_snd (x : X q) : (bit0 x).2 = bit0 x.2 := rfl
+@[simp] lemma bit1_fst (x : X q) : (bit1 x).1 = bit1 x.1 := rfl
+@[simp] lemma bit1_snd (x : X q) : (bit1 x).2 = bit0 x.2 := by { dsimp [bit1], simp, }
+
+instance : monoid (X q) :=
+{ mul_assoc := λ x y z, by { ext; { dsimp, ring }, },
+  one := ⟨1,0⟩,
+  one_mul := λ x, by { ext; simp, },
+  mul_one := λ x, by { ext; simp, },
+  ..(infer_instance : has_mul (X q)) }
+
+lemma left_distrib (x y z : X q) : x * (y + z) = x * y + x * z :=
+by { ext; { dsimp, ring }, }
+
+lemma right_distrib (x y z : X q) : (x + y) * z = x * z + y * z :=
+by { ext; { dsimp, ring }, }
+
+instance : ring (X q) :=
+{ left_distrib := left_distrib,
+  right_distrib := right_distrib,
+  ..(infer_instance : add_comm_group (X q)),
+  ..(infer_instance : monoid (X q)) }
+
+instance : comm_ring (X q) :=
+{ mul_comm := λ x y, by { ext; { dsimp, ring }, },
+  ..(infer_instance : ring (X q))}
+
+instance [fact (1 < (q : ℕ))] : nonzero_comm_ring (X q) :=
+{ zero_ne_one := λ h, begin injection h, exact @zero_ne_one (zmod q) _ h_1, end,
+  ..(infer_instance : comm_ring (X q)) }
+
+@[simp]
+lemma nat_coe_fst (n : ℕ) : (n : X q).fst = (n : zmod q) :=
+begin
+  induction n,
+  { refl, },
+  { dsimp, simp only [add_left_inj], exact n_ih, }
+end
+@[simp]
+lemma nat_coe_snd (n : ℕ) : (n : X q).snd = (0 : zmod q) :=
+begin
+  induction n,
+  { refl, },
+  { dsimp, simp only [add_zero], exact n_ih, }
+end
+
+@[simp]
+lemma int_coe_fst (n : ℤ) : (n : X q).fst = (n : zmod q) :=
+by { induction n; simp, }
+@[simp]
+lemma int_coe_snd (n : ℤ) : (n : X q).snd = (0 : zmod q) :=
+by { induction n; simp, }
+
+@[norm_cast]
+lemma coe_mul (n m : ℤ) : ((n * m : ℤ) : X q) = (n : X q) * (m : X q) :=
+by { ext; simp; ring }
+
+@[norm_cast]
+lemma coe_nat (n : ℕ) : ((n : ℤ) : X q) = (n : X q) :=
+by { ext; simp, }
+
+/-- The cardinality of `X` is `q^2`. -/
+lemma X_card : fintype.card (X q) = q^2 :=
+begin
+  dsimp [X],
+  rw [fintype.card_prod, zmod.card q],
+  ring,
+end
+
+/-- There are strictly fewer than `q^2` units, since `0` is not a unit. -/
+lemma units_card (w : 1 < q) : fintype.card (units (X q)) < q^2 :=
+begin
+  haveI : fact (1 < (q : ℕ)) := w,
+  convert card_units_lt (X q),
+  rw X_card,
+end
+
+/-- We define `ω = 2 + √3`. -/
+def ω : X q := (2, 1)
+/-- We define `ωb = 2 - √3`, which is the inverse of `ω`. -/
+def ωb : X q := (2, -1)
+
+lemma ω_mul_ωb (q : ℕ+): (ω : X q) * ωb = 1 :=
+begin
+  dsimp [ω, ωb],
+  ext; simp; ring,
+end
+
+lemma ωb_mul_ω (q : ℕ+): (ωb : X q) * ω = 1 :=
```

Comment:
```suggestion
lemma ωb_mul_ω (q : ℕ+) : (ωb : X q) * ω = 1 :=
```

Reply:
Thanks, @bryangingechen, for the suggestions. I've committed them all now.